### PR TITLE
experimental welder fuel regeneration

### LINKED ISSUE #15430: Expiremental welder does not have fuel regen.
## Description

What is said on the tin.

**Reproduction**

First use experimental welder, then wait.

### PR DESCRIPTION
**Pull request by deltanedas**
## About the PR
- adds solution regeneration component/system
- fixes #15430 by regenerating 1u of fuel every second for the experimental welding tool

stats are arbitrary but seem fine from 1 minute of testing

**Media**

https://user-images.githubusercontent.com/39013340/232413104-e99ddef5-51ac-4bf9-868d-ac2ae9327ce5.mp4

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**
:cl:
- fix: NanoTrasen lost a lawsuit regarding the false claims of fuel regeneration, experimental welding tools now regenerate 1u of fuel every second.

### CODE REVIEW

File: Content.Server/Chemistry/Components/SolutionRegenerationComponent.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    /// The reagent to be regenerated in the solution.
+    /// </summary>
+    [DataField("reagent", required: true, customTypeSerializer: typeof(PrototypeIdSerializer<ReagentPrototype>)), ViewVariables(VVAccess.ReadWrite)]
+    public string Reagent = string.Empty;
+
+    /// <summary>
+    /// Quantity of the reagent regenerated every time.
+    /// </summary>
+    [DataField("quantity"), ViewVariables(VVAccess.ReadWrite)]
+    public FixedPoint2 Quantity = FixedPoint2.New(1f);
```
Comment (0x6273) [inline Content.Server/Chemistry/Components/SolutionRegenerationComponent.cs:32]: Can this be `Solution` instead, so it's not limited to a single reagent?

Comment (deltanedas) [reply Content.Server/Chemistry/Components/SolutionRegenerationComponent.cs:32]: ill see how it gets serialized and try change it yeah

**Review by EmoGarbage404 — APPROVED** [review]
it's... it's... IT'S AMAZING

File: Content.Server/Chemistry/Components/SolutionRegenerationComponent.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    /// The name of the solution to add to.
+    /// </summary>
+    [DataField("solution", required: true), ViewVariables(VVAccess.ReadWrite)]
+    public string Solution = string.Empty;
+
+    /// <summary>
+    /// The reagent(s) to be regenerated in the solution.
+    /// </summary>
+    [DataField("added", required: true), ViewVariables(VVAccess.ReadWrite)]
+    public Solution Added = default!;
```
Comment (keronshb) [inline Content.Server/Chemistry/Components/SolutionRegenerationComponent.cs:26]: This should be renamed to "generated" or something similar.

File: Resources/Prototypes/Entities/Objects/Tools/welders.yml [hunk 3]
```diff
@@ -127,6 +127,12 @@
     enabled: false
     radius: 1.5
     color: lightblue
+  - type: SolutionRegeneration
+    solution: Welder
+    added:
```
Comment (keronshb) [inline Resources/Prototypes/Entities/Objects/Tools/welders.yml:132]: Same for here when it's changed.

**Review by keronshb — CHANGES_REQUESTED** [review]

**Review by keronshb — APPROVED** [review]
Looks good to me.

**Comment by Myakot** [comment]
This just opened up so many opportunities of copy->paste.

I wish I was more confident with git to:
1) make sinks replenishable (0.5u should be fine)
2) make water fountains with drink component
3) make atmospheric condenser shitcode (it won't actually use physics - just consume gases to auto produce water)

**Pull request merged**